test/hotspot/jtreg/testlibrary_tests/ir_framework/tests/TestScenariosCrossProduct.java

```diff
@@ -0,0 +1,123 @@
+ * @run driver ir_framework.tests.TestScenariosCrossProduct
+ */
+
+public class TestScenariosCrossProduct {
+    static void hasNFailures(String s, int count) {
+        if (!s.matches("The following scenarios have failed: (#[0-9](, )?){" + count + "}. Please check stderr for more information.")) {
+            throw new RuntimeException("Expected " + count + " failures in \"" + s + "\"");
+        }
+    }
+
+    public static void main(String[] args) {
+        // Single set should test all flags in the set by themselves.
+        try {
+            TestFramework t1 = new TestFramework();
+            t1.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=51",
+                                               "-XX:TLABRefillWasteFraction=53",
+                                               "-XX:TLABRefillWasteFraction=64"));
+            t1.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            hasNFailures(e.getMessage(), 3);
+        }
+
+        // The cross product of a set with one element and a set with three elements is three sets.
+        try {
+            TestFramework t2 = new TestFramework();
+            t2.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=53"),
+                                        Set.of("-XX:+UseNewCode", "-XX:+UseNewCode2", "-XX:+UseNewCode3"));
+            t2.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            hasNFailures(e.getMessage(), 3);
+        }
+
+        // The cross product of two sets with two elements is four sets.
+        try {
+            TestFramework t3 = new TestFramework();
+            t3.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=53", "-XX:TLABRefillWasteFraction=64"),
+                                        Set.of("-XX:+UseNewCode", "-XX:-UseNewCode"));
+            t3.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            hasNFailures(e.getMessage(), 4);
+        }
+
+        try {
+            TestFramework t4 = new TestFramework();
+            t4.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=50 -XX:+UseNewCode", "-XX:TLABRefillWasteFraction=40"),
+                                        Set.of("-XX:+UseNewCode2"));
+            t4.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            hasNFailures(e.getMessage(), 1);
+        }
+
+        // Test with an empty string. All 6 scenarios fail because 64 is the default value for TLABRefillWasteFraction.
+        try {
+            TestFramework t5 = new TestFramework();
+            t5.addCrossProductScenarios(Set.of("", "-XX:TLABRefillWasteFraction=51", "-XX:TLABRefillWasteFraction=53"),
+                                        Set.of("-XX:+UseNewCode", "-XX:+UseNewCode2"));
```

Comment:
Now looking at the implementation of `addCrossProductScenarios`: what does it do when it is called without arguments/empty args array? Can you also add a test for that?